File: 2nd_quarter/week05/whaeun/무인도 여행.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int dx[4] = {0, 1, -1, 0};
int dy[4] = {1, 0, 0, -1};

int visited[110][110];

int bfs(int x, int y, vector<string> maps){
    int ans = 0;
    queue<pair<int, int>> q;
    
    if(visited[y][x] == 1){
        return 0;
    }
    
    q.push({x, y});
    ans += maps[y][x] - '0';
    visited[y][x] = 1;
    
    while(!q.empty()){
        int p_x = q.front().first;
        int p_y = q.front().second;
        q.pop();
        
        for(int i = 0; i < 4; i++){
            int xx = dx[i] + p_x;
            int yy = dy[i] + p_y;
            
            if(xx < 0 || yy < 0 || xx >= maps[0].size() || yy >= maps.size() || visited[yy][xx] == 1 || maps[yy][xx] == 'X'){
                continue;
            }
            
            visited[yy][xx] = 1;
            q.push({xx, yy});
            ans += maps[yy][xx] - '0';
        }
    }
    
    return ans;
}

vector<int> solution(vector<string> maps) {
    vector<int> answer;
    
    vector<pair<int, int>> island;
    
    for(int i = 0; i < maps.size(); i++){
        for(int j = 0; j < maps[0].size(); j++){
            if(maps[i][j] != 'X'){
                island.push_back({j, i});
            }
        }
    }
    
    int result = 0;
    
    for(int i = 0; i < island.size(); i++){
        result = bfs(island[i].first, island[i].second, maps);
        
        if(result != 0){
            answer.push_back(result);
        }
    }

    
    sort(answer.begin(), answer.end());
    
    if(answer.size() == 0){
        answer.push_back(-1);
    }
    
    return answer;
}

File: 2nd_quarter/week05/rash/무인도여행.py
def solution(maps):
    answer = []
    visited = [[False] * len(maps[i]) for i in range(len(maps))]

    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if visited[i][j] == False and maps[i][j] != "X":
                answer.append(bfs(visited, i, j, maps))

    answer.sort()
    if len(answer) == 0:
        answer.append(-1)

    return answer


def bfs(visited, i, j, maps):
    dX = [1, 0, -1, 0]
    dY = [0, 1, 0, -1]
    queue = [[i, j]]
    visited[i][j] = True
    cnt = 0
    while queue:

        x, y = queue.pop(-1)
        cnt += int(maps[x][y])
        for k in range(4):
            nX = x + dX[k]
            nY = y + dY[k]
            if nX < 0 or nX >= len(maps) or nY < 0 or nY >= len(maps[0]) or visited[nX][nY] or maps[nX][nY] == "X":
                continue
            queue.append([nX, nY])
            visited[nX][nY] = True

    return cnt